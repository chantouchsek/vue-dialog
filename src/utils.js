export function destroyVueElement (vm) {
  if (vm && !vm._isDestroyed && (typeof vm.$destroy === 'function')) {
    vm.$destroy()
  }
}

export function findContainer (container) {
  let found
  if (typeof container === 'string') {
    found = document.querySelector(container)
  } else {
    found = container
  }
  if (!found) {
    found = document.body
  }
  return found
}

// todo
// export function middlewareSeries (promises, appContext) {
//   if (!promises.length || appContext._redirected || appContext._errored) {
//     return Promise.resolve()
//   }
//   return promisify(promises[0], appContext)
//     .then(() => {
//       return middlewareSeries(promises.slice(1), appContext)
//     })
// }
