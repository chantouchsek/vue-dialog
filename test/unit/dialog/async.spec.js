import Dialog from '../../../src/dialog'
import Vue from 'vue'
import { mount } from '../../utils'

const context = { foo: 'bar' }

describe('dialog asyncData and fetch', () => {
  beforeAll(() => {
    Dialog.prototype.context = context
  })

  test('Check asyncData', async () => {
    const asyncData = jest.fn((ctx) => {
      return new Promise((resolve) => {
        // check valid context in asyncData
        expect(ctx).toEqual({
          ...context,
          params: { bar: 'baz' }
        })
        setTimeout(() => resolve({ msg: 'foo' }), 1)
      })
    })

    const { dialog, element } = await mount({
      template: '<p>{{ msg }}</p>',
      asyncData
    }, { bar: 'baz' })

    expect(asyncData).toHaveBeenCalledTimes(1)
    expect(dialog.vmd.$data).toHaveProperty('msg', 'foo')
    expect(element.innerHTML).toMatchSnapshot()
    dialog.close()
  })

  test('Check fetch data', async () => {
    let fetched = false
    const fetch = jest.fn((ctx) => {
      return new Promise(resolve => {
        // check valid context in fetch
        expect(ctx).toEqual({
          ...context,
          params: { bar: 'baz' }
        })
        setTimeout(() => {
          fetched = true
          resolve()
        }, 1)
      })
    })

    const { dialog } = await mount({
      render: () => {},
      fetch
    }, { bar: 'baz' })

    expect(fetched).toBe(true)
    expect(fetch).toHaveBeenCalledTimes(1)
    dialog.close()
  })

  test('Check asyncData await', async () => {
    const asyncData = () => {
      return new Promise((resolve) => {
        setTimeout(() => resolve({ msg: 'foo' }), 1)
      })
    }

    const { dialog, element } = await mount({
      template: '<p>{{ msg }}</p>',
      async asyncData () {
        const data = await asyncData()
        return data
      }
    })

    expect(element.innerHTML).toMatchSnapshot()
    dialog.close()
  })

  // test('Check asyncData callback', async () => {
  //   const { dialog, element } = await mount({
  //     template: '<p>{{ msg }}</p>',
  //     asyncData (_, callback) {
  //       setTimeout(() => {
  //         callback(null, { msg: 'foo' })
  //       }, 1)
  //     }
  //   })
  //   expect(element.innerHTML).toBe('<p>foo</p>')
  //   dialog.close()
  // })

  test('Check asyncData error', async () => {
    const dlg = new Dialog({
      render: () => {},
      asyncData () {
        throw new Error('Error in asyncData')
      }
    })
    expect(dlg.show()).rejects.toThrowError('Error in asyncData')
    expect(dlg.element).toBe(null)
    expect(dlg.vm).toBe(null)
  })

  test('Check many times usage', async () => {
    const fnWait = jest.fn()

    const asyncData = jest.fn(() => {
      return new Promise((resolve) => {
        setTimeout(() => resolve({ msg: 'foo' }), 1)
      })
    })

    const fetch = jest.fn(() => {
      return new Promise((resolve) => {
        setTimeout(() => resolve(), 1)
      })
    })

    const dataFn = jest.fn(() => {
      return {
        foo: 'bar'
      }
    })

    const dlg = new Dialog({
      template: '<p>{{ msg }} {{ foo }}</p>',
      asyncData,
      fetch,
      data: dataFn
    })
    await dlg.show()
    expect(asyncData).toHaveBeenCalledTimes(1)
    expect(fetch).toHaveBeenCalledTimes(1)
    dlg.wait().then(fnWait)
    expect(dlg.element.innerHTML).toMatchSnapshot()
    dlg.close()
    await Vue.nextTick()
    expect(fnWait).toHaveBeenCalledTimes(1)

    await dlg.show()
    dlg.wait().then(fnWait)
    dlg.wait().then(fnWait)
    expect(asyncData).toHaveBeenCalledTimes(2)
    expect(fetch).toHaveBeenCalledTimes(2)
    expect(dlg.element.innerHTML).toMatchSnapshot()
    dlg.close()
    await Vue.nextTick()
    expect(fnWait).toHaveBeenCalledTimes(3)

    await dlg.show()
    expect(asyncData).toHaveBeenCalledTimes(3)
    expect(fetch).toHaveBeenCalledTimes(3)
    dlg.wait().then(fnWait)
    expect(dlg.element.innerHTML).toMatchSnapshot()
    dlg.close()
    await Vue.nextTick()
    expect(fnWait).toHaveBeenCalledTimes(4)
    expect(document.body.innerHTML).toBe('')
  })

  // test('Check asyncData error', async () => {
  //   const dlg = new Dialog({
  //     render: () => {},
  //     asyncData () {
  //       throw new Error('Error in asyncData')
  //     }
  //   })
  //   expect(dlg.show()).rejects.toThrowError('Error in asyncData')
  //   expect(dlg.element).toBe(null)
  //   expect(dlg.vm).toBe(null)
  // })
  const genCallback = (ret, timeout) => {
    return () => {
      return new Promise((resolve) => {
        setTimeout(() => resolve(ret), timeout)
      })
    }
  }

  test('Check when asyncData is object of Promises', async () => {
    var start = new Date().getTime()
    const { dialog } = await mount({
      template: '<p>{{ $data }}</p>',
      asyncData () {
        return {
          dataObject: genCallback({ test: 1 }, 10),
          dataString: genCallback('string', 30),
          dataNumber: genCallback(123, 30),
          simpleString: 'some string',
          simpleObject: { test: 123456 }
          // dataPromise: getCallback(new Promise(), 30)
        }
      }
    })
    var end = new Date().getTime()
    expect(end - start).toBeLessThan(50)
    expect(dialog.element.innerHTML).toMatchSnapshot()
    dialog.close()
  })

  test('Check when asyncData is promise of Promises', async () => {
    // var start = new Date().getTime()
    // {{ dataObject.test }}:{{ dataString }}:{{ dataNumber }}
    const { dialog } = await mount({
      template: '<p>{{ $data }}</p>',
      async asyncData () {
        return {
          dataObject: genCallback({ test: 1 }, 10),
          dataString: genCallback('string', 30),
          dataNumber: genCallback(123, 30),
          simpleString: 'some string',
          simpleObject: { test: 123456 }
          // dataPromise: getCallback(new Promise(), 30)
        }
      }
    })
    // var end = new Date().getTime()
    // expect(end - start).toBeLessThan(50)
    expect(dialog.element.innerHTML).toMatchSnapshot()
    dialog.close()
  })

  // test with errors
})
