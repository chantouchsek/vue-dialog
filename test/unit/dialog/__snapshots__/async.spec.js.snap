// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialog asyncData and fetch Check asyncData 1`] = `"<p>foo</p>"`;

exports[`dialog asyncData and fetch Check asyncData await 1`] = `"<p>foo</p>"`;

exports[`dialog asyncData and fetch Check many times usage 1`] = `"<p>foo bar</p>"`;

exports[`dialog asyncData and fetch Check many times usage 2`] = `"<p>foo bar</p>"`;

exports[`dialog asyncData and fetch Check many times usage 3`] = `"<p>foo bar</p>"`;

exports[`dialog asyncData and fetch Check when asyncData is object of Promises 1`] = `
"<p>{
  \\"simpleString\\": \\"some string\\",
  \\"simpleObject\\": {
    \\"test\\": 123456
  },
  \\"dataObject\\": {
    \\"test\\": 1
  },
  \\"dataString\\": \\"string\\",
  \\"dataNumber\\": 123
}</p>"
`;

exports[`dialog asyncData and fetch Check when asyncData is promise of Promises 1`] = `
"<p>{
  \\"simpleString\\": \\"some string\\",
  \\"simpleObject\\": {
    \\"test\\": 123456
  },
  \\"dataObject\\": {
    \\"test\\": 1
  },
  \\"dataString\\": \\"string\\",
  \\"dataNumber\\": 123
}</p>"
`;
